ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java

```diff
@@ -0,0 +1,379 @@
+      Long forkBlockNumber = iterator.next();
+      if (highestKnownFork < forkBlockNumber) {
+        highestKnownFork = forkBlockNumber;
+      }
+      if (first) {
+        // first fork
+        first = false;
+        forkList.add(
+            new ForkId(updateCrc(this.genesisHash.getHexString()), forkBlockNumber)); // Genesis
+        updateCrc(forkBlockNumber);
+
+      } else if (!iterator.hasNext()) {
+        // most recent fork
+        forkList.add(new ForkId(getCurrentCrcHash(), forkBlockNumber));
+        updateCrc(forkBlockNumber);
+        // next fork or no future fork known
+        if (currentHead > forkBlockNumber) {
+          lastKnownEntry = new ForkId(getCurrentCrcHash(), 0L);
+          forkList.add(lastKnownEntry);
+          this.forkNext = 0L;
+        } else {
+          lastKnownEntry = new ForkId(getCurrentCrcHash(), forkBlockNumber);
+          forkList.add(lastKnownEntry);
+          this.forkNext = forkBlockNumber;
+        }
+
+      } else {
+        forkList.add(new ForkId(getCurrentCrcHash(), forkBlockNumber));
+        updateCrc(forkBlockNumber);
+      }
+    }
+    return forkList;
+  }
+
+  private void updateCrc(final Long block) {
+    byte[] byteRepresentationFork = longToBigEndian(block);
+    crc.update(byteRepresentationFork, 0, byteRepresentationFork.length);
+  }
+
+  private BytesValue updateCrc(final String hash) {
+    BytesValue bv = BytesValue.fromHexString(hash);
+    byte[] byteRepresentation = bv.extractArray();
+    crc.update(byteRepresentation, 0, byteRepresentation.length);
+    return getCurrentCrcHash();
+  }
+
+  private BytesValue getCurrentCrcHash() {
+    return BytesValues.ofUnsignedInt(crc.getValue());
+  }
+
+  public static class ForkId {
+    BytesValue hash;
+    BytesValue next;
+    BytesValue forkIdRLP;
+
+    ForkId(final BytesValue hash, final BytesValue next) {
+      this.hash = hash;
+      this.next = next;
+      createForkIdRLP();
+    }
+
+    ForkId(final String hash, final String next) {
+      this.hash = BytesValue.fromHexString((hash.length() % 2 == 0 ? "" : "0") + hash);
+      if (this.hash.size() < 4) {
+        this.hash = padToEightBytes(this.hash);
+      }
+      if (next.equals("") || next.equals("0x")) {
+        this.next = BytesValue.EMPTY;
+      } else if (next.startsWith("0x")) {
+        long asLong = Long.parseLong(next.replaceFirst("0x", ""), 16);
+        this.next = BytesValues.trimLeadingZeros(BytesValue.wrap(longToBigEndian(asLong)));
+      } else {
+        this.next = BytesValue.wrap(longToBigEndian(Long.parseLong(next)));
+      }
+      createForkIdRLP();
+    }
+
+    ForkId(final String hash, final long next) {
+      this.hash = BytesValue.fromHexString(hash);
+      this.next = BytesValue.wrap(longToBigEndian(next));
+      createForkIdRLP();
+    }
+
+    ForkId(final BytesValue hash, final long next) {
+      this.hash = hash;
+      this.next = BytesValue.wrap(longToBigEndian(next));
+      createForkIdRLP();
```

Comment:
Can we reduce the duplication here too? Also, the utils you're looking for are probably in `Quantity.java`. If there are any other `0x` format related things you added, they should probably go there too.

Reply:
I was getting a circular dependency error when trying to bring this in from where Quantity.java currently is, but I may be missing something as to another means of importing it/bringing it in as a dependency.